typing/typedecl.ml

```diff
@@ -111,15 +111,19 @@ let enter_type rec_flag env sdecl (id, uid) =
   in
   let arity = List.length sdecl.ptype_params in
   if not needed then env else
+  let abstract_reason, type_manifest =
+    match sdecl.ptype_manifest, abstract_abbrevs with
+    | None, _             -> Abstract_def, None
+    | Some _, None        -> Abstract_def, Some (Btype.newgenvar ())
+    | Some _, Some reason -> reason, None
+  in
```

Comment:
My idea was that this is more versatile: one could eventually build abstract environments with other reasons.
But if you think this is an overgeneralization, I can go back to the original approach of using a boolean for `abstract_abbrevs`, and hard-coding the use of `Abstract_rec_check_regularity` here.
Or do you just mean that we should pass a value of type `abstract_reason option` at call sites ?

Reply:
After reading the code again, I understand the idea better. Some documentation of the `abstract_abbrevs` parameter would have helped.

My understanding of the intended behavior: `enter_type` will turn datatype declarations into abstract types, and type synonyms into "dummy type synonym" of the form `type t = <newgenvar()>`. If `abstract_abbrevs` is provided, then all types (datatypes or synonyms) are made abstract with the provided reason.

Documenting this intention could be nice, because the code is not so clear. I also wonder what is the purpose of those 'dummy type synonyms" -- why not just turn type synonyms into abstract types as datatype declarations? (Ideally this could be explained in a comment.)

Reply:
I added some explanations.